torch/quantization/fuse_modules.py

```diff
@@ -52,44 +61,68 @@ def fuse_conv_bn_relu(conv, bn, relu):
         return torch_fused.ConvReLU2d(
             torch.nn.utils.fusion.fuse_conv_bn_eval(conv, bn), relu)
 
+def _get_module(model, submodule_key):
+    submodule = ''
+    tokens = submodule_key.split('.')
+    cur_mod = model
+    for s in tokens:
+        if s.isdigit():
+            cur_mod = cur_mod[int(s)]
+        else:
+            cur_mod = getattr(cur_mod, s)
+    return cur_mod
+
+# Generalization of setattr
+def _set_module(model, submodule_key, module):
+    submodule = ''
+    tokens = submodule_key.split('.')
+    sub_tokens = tokens[:-1]
+    cur_mod = model
+    prev_mod = model
+    for s in sub_tokens:
+        if s.isdigit():
+            cur_mod = cur_mod[int(s)]
+        else:
+            cur_mod = getattr( cur_mod, s)
+
+    if tokens[-1].isdigit():
+        cur_mod[int(tokens[-1])] = module
+    else:
+        setattr(cur_mod, tokens[-1], module)
 
-def _fuse_modules(model, named_module_dict, modules_to_fuse, fuser_func=None):
-    assert(len(modules_to_fuse) == 2 or len(modules_to_fuse) == 3),\
-        "Can fuse only 2 or 3 modules."
+def _fuse_modules(model, modules_to_fuse, fuser_func=None):
 
     OP_LIST_TO_FUSER_FUNC = {
         (torch.nn.Conv2d, torch.nn.BatchNorm2d): fuse_conv_bn,
-        (torch.nn.Conv2d, torch.nn.BatchNorm2d, torch.nn.ReLU): fuse_conv_bn_relu
+        (torch.nn.Conv2d, torch.nn.BatchNorm2d, torch.nn.ReLU): fuse_conv_bn_relu,
+        (torch.nn.Conv2d, torch.nn.ReLU): fuse_conv_relu,
+        (torch.nn.Linear, torch.nn.ReLU): fuse_linear_relu
     }
 
     mod = []
-    parent_mod = []
-    for i in range(len(modules_to_fuse)):
-        parent_module_name = '.'.join(modules_to_fuse[i].split('.')[:-1])
-        mod.append(named_module_dict[modules_to_fuse[i]])
-        parent_mod.append(named_module_dict.get(parent_module_name, model))
+    for item in modules_to_fuse:
+        mod.append(_get_module(model, item))
 
-    new_mod = mod[0]
     if fuser_func is None:
         types = tuple(type(m) for m in mod)
         fuser_func = OP_LIST_TO_FUSER_FUNC.get(types, None)
         if fuser_func is None:
             raise NotImplementedError("Cannot fuse modules: {}".format(types))
     new_mod = fuser_func(*mod)
 
-    # Assign new_mod to module and set remaining modules to identity
-    if new_mod is not mod[0]:
-        setattr(parent_mod[0], modules_to_fuse[0].split('.')[-1], new_mod)
-        for i in range(1, len(modules_to_fuse)):
-            setattr(parent_mod[i], modules_to_fuse[i].split('.')[-1], torch.nn.Identity())
+    _set_module(model, modules_to_fuse[0], new_mod)
 
+    for item in modules_to_fuse[1:]:
+        _set_module(model, item, torch.nn.Identity())
 
 def fuse_modules(model, modules_to_fuse):
```

Comment:
I wonder whether we should do special handling for Sequential. The good thing about sequantials is that we **know** what forward looks like. So we could just iterate through them and try to find a match.

I'm also fine doing it later in a separate PR as it doesn't appear to be a critical feature.

Reply:
This is useful only when we have a sequential module at the leaf of the module tree having fusable modules, so lets do this in a later PR